### PR TITLE
net/portmapper: add pcp portmapping

### PR DESCRIPTION
**Pull request by JulianKnodt**
This adds PCP portmapping, moving all existing PCP content into a separate file, and parsing the
response from PCP portmapping requests. It hooks into the existing way we handle PMP mapping.

Tested this out on Denton's router, and seems to work fine, but would appreciate others also checking it out.
Also seems to work on pfSense on Brad's DO instance.

Updates #2563

### CODE REVIEW

File: net/portmapper/portmapper.go [hunk 1]
```diff
@@ -652,75 +669,4 @@ func (c *Client) Probe(ctx context.Context) (res ProbeResult, err error) {
...
-}
-
-type pcpResponse struct {
-	OpCode     uint8
-	ResultCode uint8
-	Lifetime   uint32
-	Epoch      uint32
-}
-
-func parsePCPResponse(b []byte) (res pcpResponse, ok bool) {
```
Comment (bradfitz) [inline net/portmapper/portmapper.go:715]: can you do the code movement in its own commit first?

File: net/portmapper/portmapper.go [hunk 2]
```diff
@@ -3,12 +3,11 @@
 // license that can be found in the LICENSE file.
 
 // Package portmapper is a UDP port mapping client. It currently allows for mapping over
-// NAT-PMP and UPnP, but will perhaps do PCP later.
+// NAT-PMP, UPnP, and PCP.
```
Comment (bradfitz) [inline net/portmapper/portmapper.go:6]: technically this belongs in the second commit :)

File: net/portmapper/portmapper.go [hunk 3]
```diff
@@ -652,75 +650,4 @@ func (c *Client) Probe(ctx context.Context) (res ProbeResult, err error) {
...
-	pkt := make([]byte, 24)
-	pkt[0] = pcpVersion // version
-	pkt[1] = pcpOpAnnounce
-	myIP16 := myIP.As16()
-	copy(pkt[8:], myIP16[:])
-	return pkt
-}
-
-// pcpMapRequest generates a PCP packet with a MAP opcode.
-func pcpMapRequest(myIP netaddr.IP, mapToLocalPort int, delete bool) []byte {
```
Comment (bradfitz) [inline net/portmapper/portmapper.go:679]: why's this being deleted? won't we need to make map requests to use PCP?

Comment (JulianKnodt) [reply net/portmapper/portmapper.go:679]: it wasn't being used previously, and I initially didn't notice it existed when I wrote `buildPCPMapRequestPacket`. I took the parts of this function that were good and moved it into that function

File: net/portmapper/portmapper.go [hunk 4]
```diff
@@ -413,6 +415,11 @@ func (c *Client) createOrGetMapping(ctx context.Context) (external netaddr.IPPor
 		return netaddr.IPPort{}, err
 	}
 
+	pcpReqMapping := buildPCPRequestMappingPacket(myIP, localPort, prevPort, pcpMapLifetimeSec)
```
Comment (bradfitz) [inline net/portmapper/portmapper.go:418]: I don't think we should race a PMP & PCP mapping. That's going to be a little chaotic.

Only send one or the other, preferring PMP for now.

File: net/portmapper/pcp.go [hunk 5]
```diff
@@ -34,41 +38,89 @@ const (
 	pcpTCPMapping = 6  // portmap TCP
 )
 
-// pcpMapRequest generates a PCP packet with a MAP opcode.
```
Comment (bradfitz) [inline net/portmapper/pcp.go]: restore comment. new version lacks a comment.

you can then document that lifetimeSec 0 means delete.

File: net/portmapper/portmapper.go [hunk 6]
```diff
@@ -432,24 +439,38 @@ func (c *Client) createOrGetMapping(ctx context.Context) (external netaddr.IPPor
...
-			}
-			if pres.OpCode == pmpOpReply|pmpOpMapUDP {
-				m.external = m.external.WithPort(pres.ExternalPort)
-				d := time.Duration(pres.MappingValidSeconds) * time.Second
-				now := time.Now()
-				m.goodUntil = now.Add(d)
-				m.renewAfter = now.Add(d / 2) // renew in half the time
-				m.epoch = pres.SecondsSinceEpoch
+			version := res[0]
+			switch version {
```
Comment (bradfitz) [inline net/portmapper/portmapper.go:443]: add a default case with a log on unexpected version?

File: net/portmapper/portmapper.go [hunk 7]
```diff
@@ -432,24 +439,38 @@ func (c *Client) createOrGetMapping(ctx context.Context) (external netaddr.IPPor
...
+					d := time.Duration(pres.MappingValidSeconds) * time.Second
+					now := time.Now()
+					m.goodUntil = now.Add(d)
+					m.renewAfter = now.Add(d / 2) // renew in half the time
+					m.epoch = pres.SecondsSinceEpoch
+				}
+			case pcpVersion:
+				pcpMapping, err := parsePCPMapResponse(res[:n])
+				if err != nil {
+					continue
```
Comment (bradfitz) [inline net/portmapper/portmapper.go:467]: log it at least? this goes nowhere now?

File: net/portmapper/portmapper.go [hunk 8]
```diff
@@ -409,10 +414,18 @@ func (c *Client) createOrGetMapping(ctx context.Context) (external netaddr.IPPor
 		}
 	}
 
-	// And ask for a mapping.
-	pmpReqMapping := buildPMPRequestMappingPacket(localPort, prevPort, pmpMapLifetimeSec)
-	if _, err := uc.WriteTo(pmpReqMapping, pmpAddru); err != nil {
-		return netaddr.IPPort{}, err
+	if haveRecentPMP {
+		// And ask for a mapping.
+		pmpReqMapping := buildPMPRequestMappingPacket(localPort, prevPort, pmpMapLifetimeSec)
```
Comment (bradfitz) [inline net/portmapper/portmapper.go:419]: the scope of this variable is two lines, so this variable can be shorter. (and be the same as the ones on line 425/426 to match)

Maybe "pkt" to match your other caller of these above?

File: net/portmapper/portmapper.go [hunk 9]
```diff
@@ -409,10 +414,18 @@ func (c *Client) createOrGetMapping(ctx context.Context) (external netaddr.IPPor
...
+		if _, err := uc.WriteTo(pmpReqMapping, pmpAddru); err != nil {
+			return netaddr.IPPort{}, err
+		}
+	} else {
+		// Only do PCP mapping in the case when PMP did not appear to be available recently.
```
Comment (bradfitz) [inline net/portmapper/portmapper.go:424]: so we send PCP unconditionally all the time even if we've never heard about its availability from an earlier probe?

I guess that's what we did for PMP previously?

Seems a little less than ideal. Maybe we should just bail out earlier if we haven't heard from anything.

File: net/portmapper/portmapper.go [hunk 10]
```diff
@@ -434,24 +447,41 @@ func (c *Client) createOrGetMapping(ctx context.Context) (external netaddr.IPPor
...
+					d := time.Duration(pres.MappingValidSeconds) * time.Second
+					now := time.Now()
+					m.goodUntil = now.Add(d)
+					m.renewAfter = now.Add(d / 2) // renew in half the time
+					m.epoch = pres.SecondsSinceEpoch
+				}
+			case pcpVersion:
+				pcpMapping, err := parsePCPMapResponse(res[:n])
+				if err != nil {
+					c.logf("Failed to get PCP mapping: %v", err)
```
Comment (bradfitz) [inline net/portmapper/portmapper.go:475]: generally style is to start log/error messages with lowercase, as they compose more nicely when there are prefixes before them.  I think vet or staticcheck or something even complains about that.

it also makes grepping logs for "fail" and "error" easier, as you don't have to do a case insensitive search

File: net/portmapper/portmapper.go [hunk 11]
```diff
@@ -434,24 +447,41 @@ func (c *Client) createOrGetMapping(ctx context.Context) (external netaddr.IPPor
...
+					continue
+				}
+				pcpMapping.internal = m.internal
+				c.mu.Lock()
+				defer c.mu.Unlock()
+				c.mapping = pcpMapping
+				return pcpMapping.external, nil
+			default:
+				c.logf("Unknown PMP/PCP version number: %d", version)
```
Comment (bradfitz) [inline net/portmapper/portmapper.go:484]: unknown (lowercase).  dump the packet?

File: net/portmapper/pcp_test.go [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+import (
+	"testing"
+)
+
+var examplePCPMapResponse = []byte{2, 129, 0, 0, 0, 0, 28, 32, 0, 2, 155, 237, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 129, 112, 9, 24, 241, 208, 251, 45, 157, 76, 10, 188, 17, 0, 0, 0, 4, 210, 4, 210, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 255, 135, 180, 175, 246}
+
+func TestParsePCPMapResponse(t *testing.T) {
+	mapping, err := parsePCPMapResponse(examplePCPMapResponse)
+	if err != nil {
+		t.Errorf("Failed to parse PCP Map Response: %v", err)
```
Comment (bradfitz) [inline net/portmapper/pcp_test.go:16]: Fatalf if the test can't proceed usefully

File: net/portmapper/pcp_test.go [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+	}
+	if err == nil && mapping == nil {
```
Comment (bradfitz) [inline net/portmapper/pcp_test.go:18]: then remove the err == nil part.

File: net/portmapper/pcp_test.go [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+		t.Errorf("Got nil mapping when expected non-nil")
```
Comment (bradfitz) [inline net/portmapper/pcp_test.go:19]: Fatalf, then see if mapping is what you expect, other than its nilness

Or reflect.DeepEqual mapping against an expected value.

File: net/portmapper/portmapper.go (hunk 9 above)
Comment (JulianKnodt) [reply net/portmapper/portmapper.go:424]: oops that's not what I intended, let me rewrite it

File: net/portmapper/portmapper.go [hunk 15]
```diff
@@ -409,10 +414,18 @@ func (c *Client) createOrGetMapping(ctx context.Context) (external netaddr.IPPor
...
-	if _, err := uc.WriteTo(pmpReqMapping, pmpAddru); err != nil {
-		return netaddr.IPPort{}, err
+	if !haveRecentPMP && haveRecentPCP {
+		// Only do PCP mapping in the case when PMP did not appear to be available recently.
+		pkt := buildPCPRequestMappingPacket(myIP, localPort, prevPort, pcpMapLifetimeSec)
+		if _, err := uc.WriteTo(pkt, pcpAddru); err != nil {
+			return netaddr.IPPort{}, err
+		}
+	} else {
+		// And ask for a mapping.
```
Comment (bradfitz) [inline net/portmapper/portmapper.go:424]: This comment style, staring with "And ask" no longer reads well in context of just the else.

File: net/portmapper/portmapper.go [hunk 16]
```diff
@@ -409,10 +414,18 @@ func (c *Client) createOrGetMapping(ctx context.Context) (external netaddr.IPPor
 		}
 	}
 
-	// And ask for a mapping.
-	pmpReqMapping := buildPMPRequestMappingPacket(localPort, prevPort, pmpMapLifetimeSec)
-	if _, err := uc.WriteTo(pmpReqMapping, pmpAddru); err != nil {
-		return netaddr.IPPort{}, err
+	if !haveRecentPMP && haveRecentPCP {
```
Comment (bradfitz) [inline net/portmapper/portmapper.go:417]: (line 412 up there where stupid github won't let me comment)  .... you're still writing to PMP to ask for the external address in the PCP case.

how do we get the external IP with PCP?

Comment (JulianKnodt) [reply net/portmapper/portmapper.go:418]: It's contained in the PCP packet, so we don't need to do anything special for it

File: net/portmapper/portmapper.go [hunk 17]
```diff
@@ -401,18 +420,31 @@ func (c *Client) createOrGetMapping(ctx context.Context) (external netaddr.IPPor
 
 	pmpAddr := netaddr.IPPortFrom(gw, pmpPort)
```
Comment (bradfitz) [inline net/portmapper/portmapper.go:421]: just make this variable be called pxpAddr and use it for both. then ditch pcpAddr{,u}

File: net/portmapper/portmapper.go [hunk 18]
```diff
@@ -401,18 +420,31 @@ func (c *Client) createOrGetMapping(ctx context.Context) (external netaddr.IPPor
...
 	pmpAddru := pmpAddr.UDPAddr()
+	pcpAddr := netaddr.IPPortFrom(gw, pcpPort)
+	pcpAddru := pcpAddr.UDPAddr()
 
-	// Ask for our external address if needed.
-	if m.external.IP().IsZero() {
-		if _, err := uc.WriteTo(pmpReqExternalAddrPacket, pmpAddru); err != nil {
+	preferPCP := DisablePMP || (!haveRecentPMP && haveRecentPCP)
```
Comment (bradfitz) [inline net/portmapper/portmapper.go:426]: shouldn't need to check DisablePMP here? If we didn't send the PMP solicitation out before, haveRecentPMP will be false, right?

But if we do need it, why don't we also need to check DisablePCP?

File: net/portmapper/portmapper.go [hunk 19]
```diff
@@ -371,23 +378,35 @@ func (c *Client) createOrGetMapping(ctx context.Context) (external netaddr.IPPor
...
+		return netaddr.IPPort{}, NoMappingError{ErrNoPortMappingServices}
+	}
+
 	// If we just did a Probe (e.g. via netchecker) but didn't
 	// find a PMP service, bail out early rather than probing
 	// again. Cuts down latency for most clients.
 	haveRecentPMP := c.sawPMPRecentlyLocked()
+	haveRecentPCP := c.sawPCPRecentlyLocked()
 
+	m := &pmpMapping{
```
Comment (bradfitz) [inline net/portmapper/portmapper.go:395]: is this used for PCP as well? if not, why is only one created up here, and why PMP?

File: net/portmapper/portmapper.go [hunk 20]
```diff
@@ -337,12 +342,18 @@ func (c *Client) createMapping() {
...
+var wildcardIP = netaddr.MustParseIP("0.0.0.0")
+
 // createOrGetMapping either creates a new mapping or returns a cached
 // valid one.
 //
 // If no mapping is available, the error will be of type
 // NoMappingError; see IsNoMappingError.
 func (c *Client) createOrGetMapping(ctx context.Context) (external netaddr.IPPort, err error) {
+	if DisableUPnP && DisablePCP && DisablePMP {
+		return
```
Comment (bradfitz) [inline net/portmapper/portmapper.go:355]: don't use naked returns on functions over 2 or 3 lines

also, why isn't this `return netaddr.IPPort{}, NoMappingError{ErrGatewayRange}` like line 359? Docs do say:

```
// If no mapping is available, the error will be of type
// NoMappingError; see IsNoMappingError.
```

Comment (JulianKnodt) [reply net/portmapper/portmapper.go:355]: Since all of them are disabled it feels odd to me that it's a `NoMappingError`, but certainly it's more right than returning nil

File: net/portmapper/portmapper.go (hunk 19 above)
Comment (JulianKnodt) [reply net/portmapper/portmapper.go:395]: this is from the old code, so I mainly left this in to preserve that, but PMP also requires two steps, getting the mapping and getting the external IP. I just shifted it around so that if it doesn't gets called it won't be allocated. I left a comment above it documenting the behaviour

File: net/portmapper/portmapper.go (hunk 18 above)
Comment (JulianKnodt) [reply net/portmapper/portmapper.go:426]: I'm not sure what you mean by PMP solicitation, other than the the request for the external IP, but I don't think that will update haveRecentPMP

I updated it to make sure that it won't try PCP if PCP is disabled
